File: sleepy-test/python-pytest/ArchiveFileNameHelper_DeleteArchiveFilesByDate_AlteredMaxArchive.py
import os
import shutil
import time
import uuid
from pathlib import Path
from pytest import fixture, mark

class TestArchiveFileNameHelper:
    def test_delete_archive_files_by_date_altered_max_archive(self):
        temp_path = Path(os.path.join(os.getenv("TEMP", "/tmp"), str(uuid.uuid4())))
        log_file = temp_path / "file.txt"
        archive_folder = temp_path / "archive"

        try:
            file_target = FileTarget(
                FileName=log_file,
                ArchiveFileName=archive_folder / "{#}.txt",
                ArchiveAboveSize=50,
                LineEnding=LineEndingMode.LF,
                ArchiveNumbering=ArchiveNumberingMode.Date,
                ArchiveDateFormat="yyyyMMddHHmmssfff",
                Layout="${message}",
                MaxArchiveFiles=5
            )
            SimpleConfigurator.ConfigureForTargetLogging(file_target, LogLevel.Debug)

            for i in range(29):
                logger.Debug("123456789")
                if i % 5 == 0:
                    time.sleep(0.05)

            LogManager.Configuration = None
            files = sorted(os.listdir(archive_folder))

            assert len(files) == file_target.MaxArchiveFiles

            file_target.MaxArchiveFiles = 2
            SimpleConfigurator.ConfigureForTargetLogging(file_target, LogLevel.Debug)
            logger.Debug("1234567890")
            LogManager.Configuration = None

            files2 = sorted(os.listdir(archive_folder))

            assert len(files2) == file_target.MaxArchiveFiles
            assert files[0] not in files2
            assert files[1] not in files2
            assert files[2] not in files2
            assert files[3] not in files2
            assert files[4] == files2[0]
            assert files2[1] not in files
        finally:
            if log_file.exists():
                log_file.unlink()
            if temp_path.exists():
                shutil.rmtree(temp_path)

File: sleepy-test/CSharpXUnit/CSharpXUnit.Tests/Tests/ArchiveFileNameHelper_DeleteArchiveFilesByDate_AlteredMaxArchive.cs
using Xunit;

namespace nlog
{
    public class ArchiveFileNameHelper
    {
        [Fact]
        public void DeleteArchiveFilesByDate_AlteredMaxArchive()
        {
            {
                var tempPath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
                var logFile = Path.Combine(tempPath, "file.txt");
                try
                {
                    string archiveFolder = Path.Combine(tempPath, "archive");
                    var fileTarget = new FileTarget
                    {
                        FileName = logFile,
                        ArchiveFileName = Path.Combine(archiveFolder, "{#}.txt"),
                        ArchiveAboveSize = 50,
                        LineEnding = LineEndingMode.LF,
                        ArchiveNumbering = ArchiveNumberingMode.Date,
                        ArchiveDateFormat = "yyyyMMddHHmmssfff", //make sure the milliseconds are set in the filename
                        Layout = "${message}",
                        MaxArchiveFiles = 5
                    };
                    SimpleConfigurator.ConfigureForTargetLogging(fileTarget, LogLevel.Debug);
                    //writing 29 times 10 bytes (9 char + linefeed) will result in 3 archive files and 1 current file
                    for (var i = 0; i < 29; ++i)
                    {
                        logger.Debug("123456789");
                        //build in a small sleep to make sure the current time is reflected in the filename
                        //do this every 5 entries
                        if (i % 5 == 0)
                            Thread.Sleep(50);
                    }
            
                    //Setting the Configuration to [null] will result in a 'Dump' of the current log entries
                    LogManager.Configuration = null;
                    var files = Directory.GetFiles(archiveFolder).OrderBy(s => s);
                    //the amount of archived files may not exceed the set 'MaxArchiveFiles'
                    Assert.Equal(fileTarget.MaxArchiveFiles, files.Count());
                    //alter the MaxArchivedFiles
                    fileTarget.MaxArchiveFiles = 2;
                    SimpleConfigurator.ConfigureForTargetLogging(fileTarget, LogLevel.Debug);
                    //writing just one line of 11 bytes will trigger the cleanup of old archived files
                    //as stated by the MaxArchiveFiles property, but will only delete the oldest files
                    logger.Debug("1234567890");
                    LogManager.Configuration = null; // Flush
                    var files2 = Directory.GetFiles(archiveFolder).OrderBy(s => s);
                    Assert.Equal(fileTarget.MaxArchiveFiles, files2.Count());
                    //the oldest files should be deleted
                    Assert.DoesNotContain(files.ElementAt(0), files2);
                    Assert.DoesNotContain(files.ElementAt(1), files2);
                    Assert.DoesNotContain(files.ElementAt(2), files2);
                    Assert.DoesNotContain(files.ElementAt(3), files2);
                    //one files should still be there
                    Assert.Equal(files.ElementAt(4), files2.ElementAt(0));
                    //one new archive file should be created
                    Assert.DoesNotContain(files2.ElementAt(1), files);
                }
                finally
                {
                    if (File.Exists(logFile))
                        File.Delete(logFile);
                    if (Directory.Exists(tempPath))
                        Directory.Delete(tempPath, true);
                }
            }
        }
    }
}
